light/beacon/api/syncer.go

```diff
@@ -0,0 +1,247 @@
+
+package api
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/light/beacon"
+	"github.com/ethereum/go-ethereum/log"
+	lru "github.com/hashicorp/golang-lru"
+)
+
+const (
+	headPollFrequency = time.Millisecond * 200
+	headPollCount     = 50
+	maxRequest        = 8
+)
+
+// CommitteeSyncer syncs committee updates and signed heads from BeaconLightApi to SyncCommitteeTracker
+type CommitteeSyncer struct {
+	api *BeaconLightApi
+
+	genesisData         beacon.GenesisData
+	checkpointPeriod    uint64
+	checkpointCommittee []byte
+	sct                 *beacon.SyncCommitteeTracker
+
+	updateCache, committeeCache *lru.Cache
+	headTriggerCh, closedCh     chan struct{}
+}
+
+// NewCommitteeSyncer creates a new CommitteeSyncer
+// Note: genesisData is only needed when light syncing (using GetInitData for bootstrap)
+func NewCommitteeSyncer(api *BeaconLightApi, genesisData beacon.GenesisData) *CommitteeSyncer {
+	updateCache, _ := lru.New(maxRequest)
+	committeeCache, _ := lru.New(maxRequest)
+	return &CommitteeSyncer{
+		api:            api,
+		genesisData:    genesisData,
+		headTriggerCh:  make(chan struct{}, 1),
+		closedCh:       make(chan struct{}),
+		updateCache:    updateCache,
+		committeeCache: committeeCache,
+	}
+}
+
+// Start starts the syncing of the given SyncCommitteeTracker
+func (cs *CommitteeSyncer) Start(sct *beacon.SyncCommitteeTracker) {
+	cs.sct = sct
+	sct.SyncWithPeer(cs, nil)
+	go cs.headPollLoop()
+}
+
+// Stop stops the syncing process
+func (cs *CommitteeSyncer) Stop() {
+	cs.sct.Disconnect(cs)
+	close(cs.headTriggerCh)
+	<-cs.closedCh
+}
+
+// headPollLoop polls the signed head update endpoint and feeds signed heads into the tracker.
+// Twice each sync period (not long before the end of the period and after the end of each period)
+// it queries the best available committee update and advertises it to the tracker.
+func (cs *CommitteeSyncer) headPollLoop() {
+	timer := time.NewTimer(0)
+	var (
+		counter      int
+		lastHead     beacon.SignedHead
+		timerStarted bool
+	)
+	nextAdvertiseSlot := uint64(1)
+
+	for {
+		select {
+		case <-timer.C:
+			timerStarted = false
+			if head, err := cs.api.GetHeadUpdate(); err == nil {
+				if !head.Equal(&lastHead) {
+					cs.updateCache.Purge()
+					cs.committeeCache.Purge()
+					if cs.sct.NextPeriod() > uint64(head.Header.Slot)>>13 {
+						cs.sct.AddSignedHeads(cs, []beacon.SignedHead{head})
+					}
+					lastHead = head
+					if uint64(head.Header.Slot) >= nextAdvertiseSlot {
+						lastPeriod := uint64(head.Header.Slot-1) >> 13
+						if cs.advertiseUpdates(lastPeriod) {
+							nextAdvertiseSlot = (lastPeriod + 1) << 13
+							if uint64(head.Header.Slot) >= nextAdvertiseSlot {
+								nextAdvertiseSlot += 8000
+							}
+						}
+					}
+				}
+				if counter < headPollCount {
+					timer.Reset(headPollFrequency)
+					timerStarted = true
+				}
+			}
+		case _, ok := <-cs.headTriggerCh:
```

Comment:
Same feeling.

Reply:
The `headPollLoop` was in a bit weird state, now it's rewritten and hopefully makes sense. Still, it's designed not just for this use case but also for light server mode where Geth is notified about new heads through the engine API. In this case it makes sense to trigger the polling externally when a new head is received and not poll the API endpoint all the time. For this purpose there is a `TriggerNewHead` function that triggers `headTriggerCh`. This function is not part of this PR because we agreed that currently unused functions should not be included in it.